python/eight_mile/tf/embeddings.py

```diff
@@ -143,234 +69,174 @@ def __init__(self, name, **kwargs):
-                char_bt_x_w = tf.reshape(self.x, [-1, mxwlen])
-                # The ablation table (4) in https://arxiv.org/pdf/1708.02182.pdf shows this has a massive impact
-                #embedding_w_dropout = tf_dropout(self.Wch, self.dropin, noise_shape=(self.vsz, 1), training=TRAIN_FLAG())
-                cembed = tf.nn.embedding_lookup(self.Wch, char_bt_x_w, name="embeddings")
-                cmot, num_filts = char_word_conv_embeddings(cembed, self.filtsz, self.dsz, self.nfeats,
-                                                            activation_fn=get_activation(self.activation),
-                                                            gating=gating_fn,
-                                                            num_gates=self.num_gates)
+        mxwlen = tf.shape(self.x)[-1]
+        char_bt_x_w = tf.reshape(self.x, [-1, mxwlen])
+        cembed = self.embed(char_bt_x_w)
+        cmot, num_filts = char_word_conv_embeddings(
+            cembed,
+            self.filtsz,
+            self.embed.output_dim,
+            self.nfeats,
+            activation_fn=get_activation(self.activation),
+            gating=gating_fn,
+            num_gates=self.num_gates
+        )
 
         if self.projsz:
-            cmot = tf.matmul(cmot, self.Wp) + self.bp
+            cmot = self.proj(cmot)
         word_char = tf.reshape(cmot, [-1, mxlen, self.outsz])
         return word_char
 
     def get_vsz(self):
-        return self.vsz
+        return self.embed.get_vsz()
 
-    # Warning this function is only initialized AFTER encode
     def get_dsz(self):
         return self.outsz
 
 
+class CharLSTMEmbeddings(TensorFlowEmbeddings):
+    def __init__(self, trainable=True, name=None, dtype=tf.float32, **kwargs):
+        trainable = kwargs.get('finetune', trainable)
+        super().__init__(trainable=trainable, name=name, dtype=dtype)
+        self.scope = kwargs.get('scope', 'CharLUT')
+        self.finetune = kwargs.get('finetune', trainable)
+        self.lstmsz = kwargs.get('lstmsz', 50)
+        self.lstm_layers = kwargs.get('layers', 1)
+        self.pdrop = kwargs.get('pdrop', 0.5)
+        self.rnn_type = kwargs.get('rnn_type', 'blstm')
+        self.x = None
+        self.embed = LookupTableEmbeddings(name=f'{self.name}/CharLUT', finetune=self.finetune, **kwargs)
+        self.lstm = BiLSTMEncoderHidden(self.embed.output_dim, self.lstmsz, self.lstm_layers, pdrop=self.pdrop, requires_length=True, name=f"{self.name}/blstm")
+
+    def encode(self, x):
+        self.x = x
+        shape = tf.shape(x)
+        B = shape[0]
+        T = shape[1]
+        W = shape[2]
+        flat_chars = tf.reshape(x, [-1, W])
+        embed_chars = self.embed(flat_chars)
+
+        # Calculate the lengths of each word
+        word_lengths = tf.reduce_sum(tf.cast(tf.equal(flat_chars, Offsets.PAD), tf.int32), axis=1)
+
+        # cuDNN throws an error if there is an input with a length of 0, this happens when the "word"
```

Comment:
A design question is if we should push this code about patching lengths of 0 into the LSTM encoder itself. When a user is using the LSTMEncoder layer I would imagine it is rare that they have a thing in the batch that is length 0 so it might be a waste of time to handle this for all LSTMEncoder calls but on the other hand this is an esoteric error so handling it under the hood would save users a lot of time if they ever hit that issue

Reply:
From discussion we decided to leave this here with comments for now and eventually push it in if other people report case where they get the error